Assets/Wizards Code/MaxArchxd/Scripts/DangerousAreaScanner.cs

```diff
@@ -0,0 +1,141 @@
+    public int detectionRadius = 10;
+
+    private NavMeshAgent agent;
+    private HashSet<int> dangerousAreasLayers;
+    private int currentArea;
+    private bool waiting;
+    private float timeSinceLastScan;
+
+    private void Awake()
+    {
+        agent = GetComponent<NavMeshAgent>();
+        dangerousAreasLayers = new HashSet<int>();
+        timeSinceLastScan = scanInterval;
+    }
+
+    void Start()
+    {
+        updateDangerousAreas();
+        currentArea = getCurrentArea();
+    }
+
+    public void updateDangerousAreas()
+    {
+        foreach (string layerName in dangerousAreas)
+        {
+            int areaIndex = NavMesh.GetAreaFromName(layerName);
+            if(areaIndex != -1) dangerousAreasLayers.Add(1 << areaIndex);
+        }
+    }
+
+    void Update()
+    {
+        if (waiting) return;
+        timeSinceLastScan += Time.deltaTime;
+        tryScan();
+    }
+
+    private void tryScan()
+    {
+        float scanDistance = scanInterval * agent.speed;
+        if (scanConditionsMet(scanDistance)) scanAreaAhead(scanDistance);
+    }
+
+    private bool scanConditionsMet(float scanDistance)
+    {
+        return timeSinceLastScan >= scanInterval && scanDistance <= agent.remainingDistance;
+    }
+
+    private void scanAreaAhead(float scanDistance)
+    {
+        NavMeshHit hit;
+        agent.SamplePathPosition(NavMesh.AllAreas, scanDistance, out hit);
+        if (hit.mask != currentArea) processAreaAhead(hit.mask);
+        timeSinceLastScan = 0;
+    }
+
+    private void processAreaAhead(int areaMaskAhead)
+    {
+        if (isAreaDangerous(areaMaskAhead)) processDangerousArea();
+        else processSafeArea(areaMaskAhead);
+    }
+
+    private bool isAreaDangerous(int areaMaskAhead)
+    {
+        return dangerousAreasLayers.Contains(areaMaskAhead) && checkForDanger();
+    }
+
+    private void processDangerousArea()
+    {
+        waiting = true;
+        agent.isStopped = true;
+        StartCoroutine("waitUntilDangerPassed");
+    }
+
+    private void processSafeArea(int areaMaskAhead)
+    {
+        currentArea = areaMaskAhead;
+        agent.isStopped = false;
+    }
+
+    private IEnumerator waitUntilDangerPassed()
+    {
+        float timeElapsed = 0;
+        while(timeElapsed < waitTime)
+        {
+            yield return null;
+            timeElapsed += Time.deltaTime;
+        }
+        waiting = false;
+    }
+
+    private int getCurrentArea()
+    {
+        NavMeshHit hit;
+
+        if(!NavMesh.SamplePosition(transform.position, out hit, 2, NavMesh.AllAreas))
+        {
+            return 0;
+        }
+        return hit.mask;
+    }
+
+    private bool checkForDanger()
+    {
+        List<NavMeshObstacle> obstacles = getRelevantObstacles();
+        foreach(NavMeshObstacle obstacle in obstacles)
+        {
+            if(Vector3.Distance(obstacle.transform.position, transform.position) <= detectionRadius) return true;
```

Comment:
This is an expensive operation if there are a reasonable number of obstacles. Using SqeMagnituted would help a little but spreading this over multiple frames would help even more.